usersession/userd/privileged_desktop_launcher.go

```diff
@@ -0,0 +1,298 @@
+	}
+
+	if filepath.Clean(desktopFile) != desktopFile {
+		return fmt.Errorf("desktop file has unclean path: %q", desktopFile)
+	}
+
+	return nil
+}
+
+// readExecCommandFromDesktopFile parses the desktop file to get the Exec entry and
+// checks that the BAMF_DESKTOP_FILE_HINT is present and refers to the desktop file.
+func readExecCommandFromDesktopFile(desktopFile string) (exec string, icon string, err error) {
+	file, err := os.Open(desktopFile)
+	if err != nil {
+		return exec, icon, err
+	}
+	defer file.Close()
+	scanner := bufio.NewScanner(file)
+
+	in_desktop_section := false
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+
+		if line == "[Desktop Entry]" {
+			in_desktop_section = true
+		} else if strings.HasPrefix(line, "[Desktop Action ") {
+			// maybe later we'll add support here
+			in_desktop_section = false
+		} else if strings.HasPrefix(line, "[") {
+			in_desktop_section = false
+		} else if in_desktop_section {
+			if strings.HasPrefix(line, "Exec=") {
+				exec = strings.TrimPrefix(line, "Exec=")
+			} else if strings.HasPrefix(line, "Icon=") {
+				icon = strings.TrimPrefix(line, "Icon=")
+			}
+		}
+	}
+
+	expectedPrefix := fmt.Sprintf("env BAMF_DESKTOP_FILE_HINT=%s "+dirs.SnapBinariesDir, desktopFile)
+	if !strings.HasPrefix(exec, expectedPrefix) {
+		return "", "", fmt.Errorf("Desktop file %q has an unsupported 'Exec' value: %q", desktopFile, exec)
+	}
+
+	return exec, icon, nil
+}
+
+// Parse the Exec command by stripping any exec variables.
+// Passing exec variables (eg, %foo) between confined snaps is unsupported. Currently,
+// we do not have support for passing them in the D-Bus API but there are security
+// implications that must be thought through regarding the influence of the launching
+// snap over the launcher wrt exec variables. For now we simply filter them out.
+// https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#exec-variables
+func parseExecCommand(exec_command string, icon string) ([]string, error) {
+	args, err := shlex.Split(exec_command)
+	if err != nil {
+		return []string{}, err
+	}
+
+	i := 0
+	for {
+		// We want to keep literal '%' (expressed as '%%') but filter our exec variables
+		// like '%foo'
+		if strings.HasPrefix(args[i], "%%") {
+			args[i] = strings.TrimPrefix(args[i], "%")
+			i++
+		} else if strings.HasPrefix(args[i], "%") {
+			switch args[i] {
+			case "%f", "%F", "%u", "%U":
+				args = append(args[:i], args[i+1:]...)
+			case "%i":
+				pre := args[:i]
+				post := args[i+1:]
+				if icon != "" {
+					post = append([]string{"--icon", icon}, post...)
+					i = i + 2
+				}
+				args = append(pre, post...)
+			default:
+				return []string{}, fmt.Errorf("cannot run %q", exec_command)
+			}
+		} else {
+			i++
+		}
+
+		if i == len(args) {
+			break
+		}
+	}
+	return args, nil
+}
+
+func (s *PrivilegedDesktopLauncher) OpenFile(parentWindow string, clientFd dbus.UnixFD, sender dbus.Sender) *dbus.Error {
```

Comment:
what's the role of this, it doesn't seem to be exposed, or called or tested? I am sure I'm missing something